keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,1338 @@
+    PodSandboxConfig sandbox_config = 3;
+}
+</pre>
+
+2. Image status request/response can also make use of the newly added runtime class field in ImageSpec to specify the runtime class being used for the image. If none is specified, containerd will assume that the request is for the default runtime class defined for that platform in containerd.toml
+
+<pre>
+message ImageStatusRequest {
+    // Spec of the image.
+    ImageSpec image = 1;
+    // Verbose indicates whether to return extra information about the image.
+    bool verbose = 2;
+}
+
+message ImageStatusResponse {
+    // Status of the image.
+    Image image = 1;
+    // Info is extra information of the Image. The key could be arbitrary string, and
+    // value should be in json format. The information could include anything useful
+    // for debug, e.g. image config for oci image based container runtime.
+    // It should only be returned non-empty when Verbose is true.
+    map<string, string> info = 2;
+}
+</pre>
+
+3. Image cleanup/kubelet garbage collection: Currently, kubelet GC will call into RemoveImage() on containerd side to remove unused images. kubelet will only pass the image name and containerd will currently delete all instances of the image. This needs to be extended on kubelet to ask for deleting (imageName, runtimeclass) to be more efficient. Optionally, we could add a behavior on containerd side to honor deletion of all versions of an image is that is was kubelet GC wants to do.
+
+### Tooling chages
+
+#### CTR changes
+
+ctr command should be extended to support pulling images per runtime class as well. 
+
+#### CRICTL changes
+
+List of crictl image commands that need to be extended:
+
+1. crictl pull image --runtime runtimeClassName imageName
+
+2. crictl images will have new column to specify the runtime class name of the image as we could now have multiple versions of the same image based on what runtime class was used to pull the image. If no runtime class was used it will specify the default runtime class name for the platform
+
+3. crictl rmi should have new option to specify runtime class name so that the appropriate image can be removed. (With this feature in place, multiple versions of the same image can exist on the node based on what runtime class was used to pull the image)
+4. crictl image info should show runtime class
+
+### Test Plan
+
+- Regression testing to ensure exisiting functionality is not regressed when no runtime class is specified while pulling the image
+- Functional/integration tests for CRI, kubelet, containerd
+
+[ ] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
```

Comment:
For Windows there are a few tests we can run, Are there any that would validate this on the Linux side?

Reply:
Yes we should have tests that test for functionality and regression across different platforms. Will call this out in the testing section.